File: modules/remotebackend/regression-tests/backend.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'json'
require 'sqlite3'

def rr(qname, qtype, content, ttl, auth = 1, domain_id = -1)
   {:qname => qname, :qtype => qtype, :content => content, :ttl => ttl.to_i, :auth => auth.to_i, :domain_id => domain_id.to_i}
end

class Handler
   def initialize(dbpath)
     @dbpath = dbpath
     @db = SQLite3::Database.new @dbpath
   end

   def db
      if block_given?
        @db.transaction
        begin
           yield @db
        rescue
           @db.rollback
           return
         end
         @db.commit
      else
         @db
      end
   end

   def do_initialize(*args)
     return true, "Test bench initialized"
   end

   def getbeforename(qname, id)
        before = db.get_first_value("SELECT ordername FROM records WHERE ordername < ? AND domain_id = ? ORDER BY ordername DESC", qname, id)
        if (before.nil?) 
           before = db.get_first_value("SELECT ordername FROM records WHERE domain_id = ? ORDER by ordername DESC LIMIT 1", id)
        end
        before
   end

  def getaftername(qname, id)
        after = db.get_first_value("SELECT ordername FROM records WHERE ordername > ? AND domain_id = ? ORDER BY ordername", qname, id)
        if (after.nil?)
           after = db.get_first_value("SELECT ordername FROM records WHERE domain_id = ? ORDER by ordername LIMIT 1", id)
        end
        after
   end


   def do_getbeforeandafternamesabsolute(args)
        args["qname"] = "" if args["qname"].nil?
        return [{:before => getbeforename(args["qname"],args["id"]), :after => getaftername(args["qname"],args["id"]), :unhashed => args["qname"]}, nil]
   end

   def do_getbeforeandafternames(args)
        args["qname"] = "" if args["qname"].nil?
        return [{:before => getbeforename(args["qname"],args["id"]), :after => getaftername(args["qname"],args["id"]), :unhashed => args["qname"]}, nil]
   end

   def do_getdomainkeys(args)
       ret = []
       db.execute("SELECT cryptokeys.id,flags,active, published, content FROM domains JOIN cryptokeys ON domains.id = cryptokeys.domain_id WHERE domains.name = ?", [args["name"]]) do |row|
          ret << {:id => row[0].to_i, :flags => row[1].to_i, :active => !(row[2].to_i.zero?), :published => row[3], :content => row[4]}
       end 
       return false if ret.empty?
       return [ret,nil]
   end

   def do_lookup(args)
     ret = []
     loop do
        begin
          sargs = {}
          if (args["zone-id"].to_i > 0)
             sargs["domain_id"] = args["zone-id"].to_i
             if (args["qtype"] == "ANY")
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND domain_id = :domain_id"
                sargs["qname"] = args["qname"]
             else
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND type = :qtype AND domain_id = :domain_id"
                sargs["qname"] = args["qname"]
                sargs["qtype"] = args["qtype"]
             end
          else
             if (args["qtype"] == "ANY")
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname"
                sargs["qname"] = args["qname"]
             else
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND type = :qtype"
                sargs["qname"] = args["qname"]
                sargs["qtype"] = args["qtype"]
             end  
          end
          db.execute(sql, sargs) do |row|
            if (row[2] == "MX" || row[2] == "SRV")
              ret << rr(row[1], row[2], row[5]+" "+row[3], row[4], row[6], row[0])
            else
              ret << rr(row[1], row[2], row[3], row[4], row[6], row[0])
            end
          end
        rescue Exception => e
          e.backtrace
          return false, [e.message]
        end
        break
     end
     return false unless ret.size > 0
     return [ret,nil]
   end
  
   def do_getdomaininfo(args) 
     ret = {}
     sql = "SELECT name,content FROM records WHERE name = :name AND type = 'SOA'"
     db.execute(sql, args) do |row|
       ret[:zone] = row[0]
       ret[:serial] = row[1].split(' ')[2].to_i
       ret[:kind] = "native"
     end
     return [ret,nil] if ret.has_key?(:zone)
     return false
   end

   def do_list(args)
     target = args["zonename"]
     ret = []
     loop do
        begin
          d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", target)
          return false if d_id.nil?
          db.execute("SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE domain_id = ?", d_id) do |row|
            if (row[2] == "MX" || row[2] == "SRV")
              ret << rr(row[1], row[2], row[5]+" "+row[3], row[4], row[6], row[0])
            else
              ret << rr(row[1], row[2], row[3], row[4], row[6], row[0])
            end
          end
        rescue Exception => e
          e.backtrace
          return false, [e.message]
        end
        break
     end
     return false unless ret.size > 0
     return [ret,nil]
   end

   def do_adddomainkey(args)
     d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
     return false if d_id.nil?
     sql = "INSERT INTO cryptokeys (domain_id, flags, active, published, content) VALUES(?,?,?,?,?)"
     active = args["key"]["active"]
     if (active) 
        active = 1
     else
        active = 0
     end
     published = args["key"]["published"]
     if (published)
         published = 1
     else
         published = 0
     end
     db do |tx|
        tx.execute(sql, [d_id, args["key"]["flags"].to_i, active, published, args["key"]["content"]])
     end
     return db.get_first_value("SELECT last_insert_rowid()").to_i
   end

   def do_deactivatedomainkey(args)
     d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
     return false if d_id.nil?
     db do |tx|
       tx.execute("UPDATE cryptokeys SET active = 0 WHERE domain_id = ? AND id = ?", [d_id, args["id"]])  
     end
     return true
   end

   def do_activatedomainkey(args)
     d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
     return false if d_id.nil?
     db do |tx|
       db.execute("UPDATE cryptokeys SET active = 1 WHERE domain_id = ? AND id = ?", [d_id, args["id"]])
     end
     return true
   end

   def do_unpublishdomainkey(args)
     d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
     return false if d_id.nil?
     db do |tx|
       tx.execute("UPDATE cryptokeys SET published = 0 WHERE domain_id = ? AND id = ?", [d_id, args["id"]])
     end
     return true
   end

   def do_publishdomainkey(args)
     d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
     return false if d_id.nil?
     db do |tx|
       db.execute("UPDATE cryptokeys SET published = 1 WHERE domain_id = ? AND id = ?", [d_id, args["id"]])
     end
     return true
   end

   def do_getdomainmetadata(args) 
	ret = []
        sql = "SELECT content FROM domainmetadata JOIN domains WHERE name = :name AND kind = :kind"
        sargs = {:name => args["name"], :kind => args["kind"]}
        db.execute(sql,sargs) do |row|
          ret << row[0]
        end
        return false unless ret.size > 0
        return [ret,nil]
   end

   def do_setdomainmetadata(args)
        d_id = db.get_first_value("SELECT id FROM domains WHERE name = ?", args["name"])
        return false if d_id.nil?
        db do |tx|
           sql = "DELETE FROM domainmetadata WHERE domain_id = ? AND kind = ?"
           tx.execute(sql, [d_id, args["kind"]])
           unless args["value"].nil?
             sql = "INSERT INTO domainmetadata (domain_id,kind,content) VALUES(?,?,?)"
             args["value"].each do |value|
               STDERR.puts"Executing INSERT INTO domainmetadata (domain_id,kind,content) VALUES(#{d_id}, #{args["kind"]}, #{value})"
               tx.execute(sql,[d_id, args["kind"], value])
             end
           end
        end
	return true
   end

   def do_directbackendcmd(args)
     return [args["query"]]
   end
end

File: modules/remotebackend/regression-tests/backend.py
#!/usr/bin/env python

import sqlite3
from pdns.remotebackend import Handler


class BackendHandler(Handler):
    def __init__(self, options={}):
        super().__init__(options=options)
        self.dbpath = options["dbpath"]
        self.db = sqlite3.connect(self.dbpath)

    def get_domain_id(self, name):
        cur = self.db.execute("SELECT id FROM domains WHERE name = ?", (name,))
        row = cur.fetchone()
        if not row:
            self.result = False
            raise KeyError
        return int(row[0])

    def record(self, qname="", qtype="", content="", ttl=1, prio=0, auth=1, domain_id=-1):
        """Generate one record"""
        if ttl == -1:
            ttl = self.ttl
        if qtype in ("MX", "SRV"):
            content = "%d %s" % (prio, content)
        return {"qtype": qtype, "qname": qname, "content": content, "ttl": ttl, "auth": auth, "domain_id": domain_id}

    # ends up here as qname=qname, id=id
    def getbeforename(self, **kwargs):
        cur = self.db.execute(
            "SELECT ordername FROM records WHERE ordername < :qname AND domain_id = :id ORDER BY ordername DESC LIMIT 1",
            kwargs,
        )
        row = cur.fetchone()
        if not row:
            cur = self.db.execute(
                "SELECT ordername FROM records WHERE domain_id = :id ORDER by ordername DESC LIMIT 1", kwargs
            )
            row = cur.fetchone()
        result = row[0]
        if row[0] is None:
            result = ""
        return result

    def getaftername(self, **kwargs):
        cur = self.db.execute(
            "SELECT ordername FROM records WHERE ordername > :qname AND domain_id = :id ORDER BY ordername LIMIT 1",
            kwargs,
        )
        row = cur.fetchone()
        if row is None:
            cur = self.db.execute(
                "SELECT ordername FROM records WHERE domain_id = :id ORDER by ordername LIMIT 1", kwargs
            )
            row = cur.fetchone()
        result = row[0]
        if row[0] is None:
            result = ""
        return result

    def do_getbeforeandafternamesabsolute(self, **kwargs):
        self.result = {
            "before": self.getbeforename(**kwargs),
            "after": self.getaftername(**kwargs),
            "unhashed": kwargs["qname"],
        }

    def do_getbeforeandafternames(self, **kwargs):
        self.do_getbeforeandafternamesabsolute(**kwargs)

    def do_getdomainkeys(self, name, **kwargs):
        self.result = []
        cur = self.db.execute(
            "SELECT cryptokeys.id, flags, active, published, content FROM domains JOIN cryptokeys ON domains.id = cryptokeys.domain_id WHERE domains.name = :name",
            {"name": name},
        )
        for row in cur.fetchall():
            self.result.append(
                {"id": row[0], "flags": row[1], "active": row[2] != 0, "published": row[3], "content": row[4]}
            )
        if len(self.result) == 0:
            self.result = False
        self.log.append(self.dbpath)

    def do_lookup(self, qname="", qtype="", domain_id=-1, **kwargs):
        self.result = []
        if kwargs.get("zone-id", -1) > 0:
            domain_id = kwargs["zone-id"]
        if domain_id > -1:
            if qtype == "ANY":
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND domain_id = :domain_id"
            else:
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND type = :qtype AND domain_id = :domain_id"
        else:
            if qtype == "ANY":
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname"
            else:
                sql = "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE name = :qname AND type = :qtype"
        cur = self.db.execute(sql, {"qname": qname, "qtype": qtype, "domain_id": domain_id})
        for row in cur.fetchall():
            self.result.append(
                self.record(
                    qname=row[1], qtype=row[2], content=row[3], ttl=row[4], prio=row[5], auth=row[6], domain_id=row[0]
                )
            )

    def do_getdomaininfo(self, name="", **kwargs):
        self.result = False
        cur = self.db.execute(
            "SELECT domain_id,name,content FROM records WHERE name = :name AND type = 'SOA'", {"name": name}
        )
        for row in cur.fetchall():
            self.result = {
                "zone": row[1],
                "serial": int(row[2].split(" ")[2]),
                "kind": "native",
                "id": row[0],
            }

    def do_getalldomains(self):
        self.result = []
        cur = self.db.execute(
            "SELECT domain_id,name,content FROM records WHERE name = :name AND type = 'SOA'", {"name": name}
        )
        for row in cur.fetchall():
            self.result.append(
                {
                    "zone": row[1],
                    "serial": int(row[2].split(" ")[2]),
                    "kind": "native",
                    "id": row[0],
                }
            )

    def do_list(self, zonename="", domain_id=-1, **kwargs):
        if domain_id == -1:
            try:
                domain_id = self.get_domain_id(zonename)
            except KeyError:
                return
        if domain_id > -1:
            self.result = []
            cur = self.db.execute(
                "SELECT domain_id,name,type,content,ttl,prio,auth FROM records WHERE domain_id = ?", (domain_id,)
            )
            for row in cur.fetchall():
                self.result.append(
                    self.record(
                        qname=row[1],
                        qtype=row[2],
                        content=row[3],
                        ttl=row[4],
                        prio=row[5],
                        auth=row[6],
                        domain_id=row[0],
                    )
                )

    def do_adddomainkey(self, name, key, **kwargs):
        try:
            domain_id = self.get_domain_id(name)
        except KeyError:
            return
        key["domain_id"] = domain_id

        cur = self.db.execute(
            "INSERT INTO cryptokeys (domain_id, flags, active, published, content) VALUES(:domain_id, :flags, :active, :published, :content)",
            key,
        )
        self.db.commit()

        self.result = cur.lastrowid
        self.log.append(self.dbpath)

    def do_deactivatedomainkey(self, **kwargs):
        try:
            domain_id = self.get_domain_id(kwargs["name"])
        except KeyError:
            return
        kwargs["domain_id"] = domain_id

        self.db.execute("UPDATE cryptokeys SET active = 0 WHERE domain_id = :domain_id AND id = :id", kwargs)
        self.db.commit()

        self.result = True

    def do_activatedomainkey(self, **kwargs):
        try:
            domain_id = self.get_domain_id(kwargs["name"])
        except KeyError:
            return
        kwargs["domain_id"] = domain_id

        self.db.execute("UPDATE cryptokeys SET active = 1 WHERE domain_id = :domain_id AND id = :id", kwargs)
        self.db.commit()

        self.result = True

    def do_unpublishdomainkey(self, **kwargs):
        try:
            domain_id = self.get_domain_id(kwargs["name"])
        except KeyError:
            return
        kwargs["domain_id"] = domain_id

        self.db.execute("UPDATE cryptokeys SET published = 0 WHERE domain_id = :domain_id AND id = :id", kwargs)
        self.db.commit()

        self.result = True

    def do_publishdomainkey(self, **kwargs):
        try:
            domain_id = self.get_domain_id(kwargs["name"])
        except KeyError:
            return
        kwargs["domain_id"] = domain_id

        self.db.execute("UPDATE cryptokeys SET published = 1 WHERE domain_id = :domain_id AND id = :id", kwargs)
        self.db.commit()

        self.result = True

    def do_getalldomainmetadata(self, name, **kwargs):
        cur = self.db.execute(
            "SELECT kind, content FROM domainmetadata JOIN domains WHERE name = :name", {"name": name}
        )
        self.result = {}
        for row in cur.fetchall():
            if not row[0] in self.result:
                self.result[row[0]] = list()
            self.result[row[0]].append(row[1])

    def do_getdomainmetadata(self, name, kind, **kwargs):
        cur = self.db.execute(
            "SELECT content FROM domainmetadata JOIN domains WHERE name = :name AND kind = :kind",
            {"name": name, "kind": kind},
        )
        self.result = cur.fetchall()

    def do_setdomainmetadata(self, name, kind, value, **kwargs):
        try:
            domain_id = self.get_domain_id(name)
        except KeyError:
            return

        self.db.execute(
            "DELETE FROM domainmetadata WHERE domain_id = :domain_id AND kind = :kind",
            {"domain_id": domain_id, "kind": kind},
        )
        if value:
            self.db.execute(
                "INSERT INTO domainmetadata (domain_id,kind,content) VALUES(:domain_id, :kind, :content)",
                {"domain_id": domain_id, "kind": kind, "content": content},
            )
        self.db.commit()

    def do_starttransaction(self, trxid, **kwargs):
        pass

    def do_committransaction(self, trxid, **kwargs):
        pass

    def do_directbackendcmd(self, query, **kwargs):
        self.result = query
